hierarchicalforecast/methods.py

```diff
@@ -440,7 +444,64 @@ def fit_predict(self,
 
     __call__ = fit_predict
 
-# %% ../nbs/methods.ipynb 32
+# %% ../nbs/methods.ipynb 25
+class TopDownSparse(TopDown):
+    """TopDownSparse Reconciliation Class.
+
+    This is an implementation of top-down reconciliation using the sparse matrix
+    approach. It works much more efficiently on data sets with many time series.
+
+    See the parent class for more details.
+    """
+
+    is_sparse_method = True
+
+    def _get_PW_matrices(
+        self,
+        S: sparse.csr_matrix,
+        y_hat: np.ndarray,
+        tags: Dict[str, np.ndarray],
+        y_insample: Optional[np.ndarray] = None,
+    ):
+        if not is_strictly_hierarchical(S, tags):
+            raise ValueError(
+                "Top down reconciliation requires strictly hierarchical structures."
+            )
+
+        n_hiers, n_bottom = S.shape
+        idx_top = int(S.sum(axis=1).argmax())
+        levels_ = dict(sorted(tags.items(), key=lambda x: len(x[1])))
+        idx_bottom = levels_[list(levels_)[-1]]
+
+        y_top = y_insample[idx_top]
+        y_btm = y_insample[idx_bottom]
+        if self.method == "average_proportions":
+            prop = np.mean(y_btm / y_top, axis=1)
+        elif self.method == "proportion_averages":
+            prop = np.mean(y_btm, axis=1) / np.mean(y_top)
+        elif self.method == "forecast_proportions":
+            raise Exception(f"Fit method not implemented for {self.method} yet.")
+        else:
+            raise Exception(f"Unknown method {self.method}.")
+
+        P = sparse.coo_matrix(
+            (
+                prop,
+                (
+                    np.arange(len(prop), dtype=np.min_scalar_type(n_bottom)),
+                    np.full_like(prop, idx_top, np.min_scalar_type(n_hiers)),
+                ),
+            ),
+            shape=(n_bottom, n_hiers),
+            dtype=np.float64,
+        ).tocsr()
```

Comment:
Isn't it more efficient to create this as a `csr_matrix` immediately to avoid the conversion time?

Reply:
That is an excellent idea, thanks. I just pushed the changes. It also allows us to quickly construct $P$ directly using the required ``indices`` and ``indptr`` to save even more time relative to constructing using the *ijv* format.